### ISSUE TITLE
Vienna Keeps Waiting for Unresponding Website

### ISSUE THREAD

**Issue by faxotherapy** [ViennaRSS/vienna-rss#1380]
Hello, Vienna does not stop after a certain time when reaching unresponding website, e.g.: https://mullvad.net/blog/feed/rss

This is a bit annoying, because this issue prevents Vienna from refreshing every x hours since Vienna keeps contacting/waiting unresponding website **indefinitely**.

Currently, I have to manually click on the the X icon to stop Vienna so it can continue checking news at a later time.

Would it be possible to tell Vienna to give up a connexion to a website when the latter is not responding.

Thank you.

![ViennaStuck](https://user-images.githubusercontent.com/6538008/100541294-d51ef580-323a-11eb-86d8-407568c803f5.gif)

**Comment by barijaona** [comment]
I am not sure I understand the issue. There is a timeout which is currently configured to 3 minutes, so a feed request should not block later requests on the same feed.

**Comment by barijaona** [comment]
Current state of #1392 is a workaround for handling a hanging connection.

But the fact that this Mullvad site is not responding to Vienna is in itself a deeper issue: neither a redirection status nor any data are received by the current version of Vienna.

I have checked that Vienna 3.4.2 (last version using NSURLConnection instead of NSURLSession for refreshing feeds) works fine with this feed…


The problem seems to be caused by a failure in negotiation of a secure connection (-1200 = NSURLErrorSecureConnectionFailed). Here is the interesting part of the log:
````
2020-12-11 04:57:25.240273+0300 Vienna[35487:3320092] Connection 1: received failure notification
2020-12-11 04:57:25.240357+0300 Vienna[35487:3320092] Connection 1: failed to connect 3:-9816, reason -1
2020-12-11 04:57:25.240396+0300 Vienna[35487:3320092] Connection 1: encountered error(3:-9816)
2020-12-11 04:58:25.692740+0300 Vienna[35487:3320896] Connection 2: received failure notification
2020-12-11 04:58:25.692799+0300 Vienna[35487:3320896] Connection 2: failed to connect 3:-9816, reason -1
2020-12-11 04:58:25.692828+0300 Vienna[35487:3320896] Connection 2: encountered error(3:-9816)
2020-12-11 04:58:26.148650+0300 Vienna[35487:3320896] [boringssl] boringssl_context_handle_fatal_alert(1763) [C3.1:2][0x7ff34190a9b0] read alert, level: fatal, description: protocol version
2020-12-11 04:58:26.148860+0300 Vienna[35487:3320896] [boringssl] boringssl_session_handshake_incomplete(90) [C3.1:2][0x7ff34190a9b0] SSL library error
2020-12-11 04:58:26.148954+0300 Vienna[35487:3320896] [boringssl] boringssl_session_handshake_error_print(41) [C3.1:2][0x7ff34190a9b0] Error: 140682753804728:error:1000042e:SSL routines:OPENSSL_internal:TLSV1_ALERT_PROTOCOL_VERSION:/AppleInternal/BuildRoot/Library/Caches/com.apple.xbs/Sources/boringssl/boringssl-351.40.2/ssl/tls_record.cc:592:SSL alert number 70
2020-12-11 04:58:26.149027+0300 Vienna[35487:3320896] [boringssl] nw_protocol_boringssl_handshake_negotiate_proceed(767) [C3.1:2][0x7ff34190a9b0] handshake failed at state 12288: not completed
2020-12-11 04:58:26.149841+0300 Vienna[35487:3320896] Connection 3: received failure notification
2020-12-11 04:58:26.149920+0300 Vienna[35487:3320896] Connection 3: failed to connect 3:-9816, reason -1
2020-12-11 04:58:26.149961+0300 Vienna[35487:3320896] Connection 3: encountered error(3:-9816)
2020-12-11 04:58:26.150373+0300 Vienna[35487:3320896] Task <2C35A68A-B9AA-49A6-80E7-F6F02DEF3BEC>.<1> HTTP load failed, 0/0 bytes (error code: -1200 [3:-9816])
````

**Comment by josh64x2** [comment]
@barijaona have you tried playing around with `TLSMinimumSupportedProtocol` as per https://developer.apple.com/documentation/foundation/nsurlsessionconfiguration/1411526-tlsminimumsupportedprotocol ?

If not, I'll give it a go later tonight.

**Comment by josh64x2** [comment]
Here are the ciphers supported according to nmap:

```
| ssl-enum-ciphers: 
|   TLSv1.2: 
|     ciphers: 
|       TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256 (ecdh_x25519) - A
|       TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 (ecdh_x25519) - A
|       TLS_DHE_RSA_WITH_AES_256_GCM_SHA384 (dh 4096) - A
|       TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 (ecdh_x25519) - A
|       TLS_DHE_RSA_WITH_AES_128_GCM_SHA256 (dh 4096) - A
|     compressors: 
|       NULL
|     cipher preference: server
|_  least strength: A

```

**Comment by barijaona** [comment]
I've unsuccessfully tested, under macOS 11 as well as 10.11, various combinations of `TLSMinimumSupportedProtocolVersion`, `TLSMaximumSupportedProtocolVersion`, `TLSMinimumSupportedProtocol` and `TLSMaximumSupportedProtocol` settings.

I also tested the mullvad.net site with Apple's [TLSTool](https://developer.apple.com/library/archive/samplecode/sc1236/Introduction/Intro.html) and did not find any clue.

**Comment by barijaona** [comment]
I can confirm that the problem is related to the NSURLSession object that we use in RefreshManager.

The connection works fine if I replace `self.urlSession` with `[NSURLSession sharedSession]` in https://github.com/ViennaRSS/vienna-rss/blob/3452c8e1274f8eb8f03bc053cec7186cfed6baea/Vienna/Sources/Fetching/RefreshManager.m#L1126

**Comment by Eitot** [comment]
The issue also seems to disappear if you comment out the two URLSession delegate methods that RefreshManager implements. The shared session does not use a delegate callback.

**Comment by barijaona** [comment]
The action for `NSURLAuthenticationMethodServerTrust` is wrong

**Comment by barijaona** [comment]
Fixed by #1392 and #1395
